modules/vectorfieldvisualization/integrallinetracer.h

```diff
@@ -32,34 +32,281 @@
+        }
+
+        integrate(stepsFWD, p, line, true);
+
+        return res;
+    }
+
+    void addMetaDataSampler(const std::string &name, std::shared_ptr<const Sampler> sampler) {
+        metaSamplers_[name] = sampler;
+    }
+
+    const DataHomogenouSpatialMatrixrix &getSeedTransformationMatrix() const {
+        return seedTransformation_;
+    }
+
+    void setTransformOutputToWorldSpace(bool transform) {
+        transformOutputToWorldSpace_ = transform;
+    }
+    bool isTransformingOutputToWorldSpace() const { return transformOutputToWorldSpace_; }
+
+private:
+    bool addPoint(IntegralLine &line, SpatialVector pos) {
+        auto worldVelocty = sampler_->sample(pos);
+
+        if (glm::length(worldVelocty) < std::numeric_limits<double>::epsilon()) {
+            return false;
+        }
+
+        if (transformOutputToWorldSpace_) {
+            SpatialVector worldPos =
+                detail::seedTransform<Sampler, DataVector, DataHomogenousVector>(
+                    typename std::integral_constant<bool, TimeDependent>::type(), toWorld_, pos);
+
+            line.getPositions().emplace_back(util::glm_convert<dvec3>(worldPos));
+        } else {
+            line.getPositions().emplace_back(util::glm_convert<dvec3>(pos));
+        }
+
+        line.getMetaData<dvec3>("velocity").emplace_back(util::glm_convert<dvec3>(worldVelocty));
+
+        if (TimeDependent) {
+            line.getMetaData<double>("timestamp").emplace_back(pos[Sampler::SpatialDimensions - 1]);
+        }
+
+        for (auto &m : metaSamplers_) {
+            line.getMetaData<typename Sampler::ReturnType>(m.first).emplace_back(
+                util::glm_convert<dvec3>(m.second->sample(pos)));
+        }
+        return true;
+    }
+
+    void integrate(size_t steps, SpatialVector pos, IntegralLine &line, bool fwd) {
+        for (int i = 0; i < steps; i++) {
+            if (!sampler_->withinBounds(pos)) {
+                if (fwd) {
+                    line.setForwardTerminationReason(IntegralLine::TerminationReason::OutOfBounds);
+                } else {
+                    line.setBackwardTerminationReason(IntegralLine::TerminationReason::OutOfBounds);
+                }
+                break;
+            }
+
+            pos = detail::step<TimeDependent, SpatialVector, DataVector>(
+                pos, integrationScheme_, stepSize_ * (fwd ? 1.0 : -1.0), invBasis_,
+                normalizeSamples_, *sampler_);
+
+            if (!addPoint(line, pos)) {
+                if (fwd) {
+                    line.setForwardTerminationReason(IntegralLine::TerminationReason::ZeroVelocity);
+                } else {
+                    line.setBackwardTerminationReason(
+                        IntegralLine::TerminationReason::ZeroVelocity);
+                }
+
+                break;
+            }
+        }
+        if (fwd) {
+            line.setForwardTerminationReason(IntegralLine::TerminationReason::Steps);
```

Comment:
how about a setTerminationReason(Direction d, Reason r) function, to avoid all this if(fwd) all over the place...

Reply:
will this not override the ones set inside of the for loop?